cmd/snap-update-ns/user.go

```diff
@@ -33,24 +35,51 @@ type UserProfileUpdateContext struct {
 	// the update operation is occurring. It may be the current UID but doesn't
 	// need to be.
 	uid int
+	// home contains the user's real home directory
+	home string
+}
+
+// isPlausibleHome returns and error if the path is not clean and absolute path to a directory
+// (symlinks not followed) that can be opened for reading by the user.
+func isPlausibleHome(path string) error {
+	if path != filepath.Clean(path) {
+		return fmt.Errorf("cannot allow unclean path")
+	}
+	if !filepath.IsAbs(path) {
+		return fmt.Errorf("cannot allow relative path")
+	}
+	const openFlags = syscall.O_NOFOLLOW | syscall.O_CLOEXEC | syscall.O_DIRECTORY
+	fd, err := sysOpen(path, openFlags, 0)
+	if err != nil {
+		return err
+	}
+	sysClose(fd)
+	return nil
 }
 
 // NewUserProfileUpdateContext returns encapsulated information for performing a per-user mount namespace update.
-func NewUserProfileUpdateContext(instanceName string, fromSnapConfine bool, uid int) *UserProfileUpdateContext {
+func NewUserProfileUpdateContext(instanceName string, fromSnapConfine bool, uid int) (*UserProfileUpdateContext, error) {
+	realHome, err := snapEnvRealHome()
+	if err != nil {
+		return nil, fmt.Errorf("cannot retrieve home directory: %v", err)
+	}
+	// See bootstrap.c function switch_to_privileged_user(). When snap-update-ns is invoked for
+	// user mounts, the effective uid and gid is changed to the calling user and supplementary
+	// groups dropped, while retaining capability SYS_ADMIN. Having the effective uid and gid
+	// changed to the calling user is a prerequisite for isPlausibleHome to function as intended.
```

Comment:
as discussed we need to clarify whether the check here just helps failing early, or if has real security implications?

Reply:
This was discussed during RIGA meeting with Samuele/Alex.
1. --usermount/user profiles are only actioned for non-root users
2. At this point all capabilities, except for SYS_ADMIN is dropped, and user is changed irreversibly to the real user and real user's supplementary groups are dropped.
3. On systems where apparmor is available, the "owner" rules that also limits access to the actual user home.

In summary,  isPlausibleHome, is only meant for best effort at failing early. If somehow this checkpoint is passed where not intended, it is still not possible to create directories in any location where the DAC access (and apparmor where when available) does not allow the real user permissions to do so.